Overhaul Docker support

terminalmage opened the pull request:
This PR does the following:

- Splits states for container/volume/image/network management into four separate state modules.
- Preserves backward compatibility by making ``docker.image_present`` invoke ``docker_image.present``, etc.
- Changes how Salt detects that a container needs to be replaced. Instead of comparing each passed argument to the named container's configuration, it creates a temporary container, and compares that container to the named container. If the two differ, then the older container is removed, and the new one is renamed and started, becoming the named container.
- Removes the unit tests for container management and replaces them with integration tests.
- Adds unit tests for the new salt.utils.docker

terminalmage committed:
Overhaul Docker support

This does the following:

- Splits states for container/volume/image/network management into four
  separate state modules.

- Preserves backward compatibility by making ``docker.image_present``
  invoke ``docker_image.present``, etc.

- Changes how Salt detects that a container needs to be replaced.
  Instead of comparing each passed argument to the named container's
  configuration, it creates a temporary container, and compares that
  container to the named container. If the two differ, then the older
  container is removed, and the new one is renamed and started, becoming
  the named container.

- Removes the unit tests for container management and replaces them with
  integration tests.

- Adds unit tests for the new salt.utils.docker

salt-jenkins commented:
@terminalmage, thanks for your PR! By analyzing the history of the files in this pull request, we identified @ticosax, @rhansen and @lwh to be potential reviewers.

rallytime commented:
Go Go Jenkins!

rallytime commented:
@terminalmage After the latest test run here, looks like there are a couple of pylint errors:

https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/9472/violations/

And then one test failure:
- `unit.modules.test_docker.DockerTestCase.test_check_mine_cache_is_refreshed_on_container_change_event`

Can you take a look at those when you get a moment?

terminalmage commented:
I've fixed the remaining lint errors locally and will tackle that unit test later today or tomorrow. The reason it's happening is because the container creation code now invokes a function I added which gets argspecs of docker-py's API functions so that we know which arguments to pull from the passed kwargs and send to the API to create the container. The function I wrote to get these argspecs just needs to be mocked for the unit test to pass. Easy fix.

terminalmage commented:
The fix was actually something different than I thought. Fix pushed, tests should pass now.

terminalmage commented:
However, we still need to check that the integration tests were not skipped.

ticosax commented:
> it creates a temporary container

I'm afraid this assumption is not universal for all containers. I have for instance some containerized processes that has to be treated like singleton, thus can't be run in parallel. eg: `celery beat`.

Anyway, I will give it a try. I will report feedback within a couple of days.

terminalmage commented:
@ticosax The important distinction is that it runs `create()` and _not_ `start()`. So the temp container is never started, unless it replaces the previous container. That should work fine for your use case.

terminalmage commented:
There are still some failures, which I spoke with @s0undt3ch about and he believes are due to another test in the suite modifying sys.path when it shouldn't be. He's working on some chanegs that he hopes will catch things like that, and this PR will need a rebase once that is done.

terminalmage commented:
Go Go Jenkins!

terminalmage commented:
In the meantime I've kicked off another round of tests to see if my integration tests pass now that we've gotten salt-jenkins properly configured to install docker and busybox.

cachedout commented:
My lag time is too high. I need this rebased. Apologies. :-/

terminalmage commented:
[This failure](https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/9551/testReport/junit/unit.modules.test_docker/DockerTestCase/test_check_mine_cache_is_refreshed_on_container_change_event/) is due to a sys.path issue that @s0undt3ch is trying to track down in the unit tests, causing imports to get messed up.

The Cent7 failures are due to me using an argument that is not supported. This is actually good news! It means that my tactic of letting the API manage what is supported and what is not, rather than tracking it in a massive dictionary like we were doing before, is working beautifully. It does however mean that I'll need to update the tests to use this option _only_ when the API version is 1.25 or greater. It also means that tests may take longer to run unless I just forcibly remove them (which probably won't be that big of a deal for these tests as the containers created by them are just running `sleep`).

I'll get the tests updated.

rallytime commented:
Sorry @terminalmage, but it looks like this needs another rebase.

cachedout commented:
@ticosax Please let me know when you have your feedback ready. Thanks!

terminalmage commented:
OK, I fixed the unit tests on Ubuntu by changing the imports in those test modules. Instead of doing ``from salt.modules import docker as docker_mod`` it now does ``import salt.modules.docker as docker_mod``.

For some reason which I don't quite grok as my knowledge of what happens on an import is not as comprehensive as it could be, the act of importing a module _from_ another module _as_ a different module name causes that module to exist in the global namespace such that doing ``from salt.modules import docker as docker_mod`` causes ``salt.utils.docker`` to import that module instead of docker-py when it executes its ``import docker``.

So, pending review, this should be good to merge.

cachedout commented on salt/states/docker_image.py:
```diff
@@ -0,0 +1,419 @@
+
+    sls
+        Allow for building images with ``dockerng.sls_build`` by specify the
+        sls files to build with. This can be a list or comma-seperated string.
```
`SLS`

cachedout commented on salt/utils/docker.py:
```diff
@@ -0,0 +1,994 @@
+        if salt.utils.is_dictlist(kwargs[key]):
+            kwargs[key] = salt.utils.repack_dictlist(kwargs[key])
+
+        if _matches(key, 'cpuset_cpus', 'cpuset_mems', 'hostname', 'ipc_mode',
```
Awfully big ifel chainghere

ticosax commented:
@cachedout I'll test it from develop branch. I don't want to block it.
Thx

cachedout commented:
@ticosax Sounds good. Thank you as always.

terminalmage commented on salt/modules/docker.py:
```diff
@@ -3068,167 +2394,171 @@ def create(image,
+        Example: ``user=foo``
 
-        Example: ``pid_mode=host``
+    userns_mode
```
I added an alias for this, as we are first and foremost using the docker-py argument names.

terminalmage commented:
Per recommendation by @cachedout I have refactored the translate_input function, placing the translation logic in a separate module to break up the massive amounts of if statements and continues.

I also found 4 arguments which I was translating input for which lack a test (!!!), so this has been rectified as well.

ticosax commented:
@terminalmage I could give it a spin.
These are great improvements you are bringing here :smile: 
I just noticed few glitches so far.

1. I have undesired diffs
 ``` container:
                  ----------
                  HostConfig:
                      ----------
                      Dns:
                          ----------
                          new:
                              None
                          old:
                      DnsOptions:
                          ----------
                          new:
                              None
                          old:
                      DnsSearch:
                          ----------
                          new:
                              None
                          old:
```

2. a container with `start: false` will return `result=False` (considered like a failure).
```diff
index a8d6c9f..1239094 100644
--- a/salt/states/docker_container.py
+++ b/salt/states/docker_container.py
@@ -1743,7 +1743,7 @@ def running(name,
         comments.append('Container has a new image')
         ret['changes']['image'] = {'old': current_image_id, 'new': image_id}
 
-    if post_state != 'running':
+    if post_state != 'running' and start:
         comments.append('Container is not running')
     else:
         ret['result'] = True
```
this patch should do it.

3. mod_watch is lacking support for `stopped` eg:
```yaml
stop-nginx:
  dockerng.stopped:
    - containers:
      - nginx
    - error_on_absent: False
    - onchanges:
      - dockerng: django-image
```

terminalmage commented:
@ticosax Thanks for giving this a look. Could you let me know what you changed in the SLS to generate that diff? I agree, that is not expected. If you did not change anything, could you please post the results of a ``docker.inspect_container <containername> --out=pprint``, as well as the SLS you used which resulted in that diff?

Good catch on that ``start`` issue. It turns out I misunderstood how it was supposed to work, and wrote bad tests to confirm my mistaken assumptions, so I'll also need to fix the tests associated with it. Can you answer the following questions about ``start=False``, so I make sure I get it right?

1. Container exists and was running already, and was replaced. Should the new container be started?
2. Container exists and was *not* running already, and was replaced. Should the new container be started?
3. Container exists but was not replaced. It should not be started, correct?
4. Container did not previously exist. It should still be started even with ``start=False``, because it is a new container, correct?

I'm pretty sure on number 3 and 4 but not sure on 1 and 2. Thanks for helping to clear this up.

Thanks again for reviewing. I'll take care of that mod_watch issue as well.

terminalmage commented:
@ticosax It turns out that the pre-existing code also did not have any support for ``stopped`` in the mod_watch, either. Just FYI.

ticosax commented:
The sls are the same, please find requested resources.
Nothing fancy I think.
sls:
```yaml
locksocket:
  dockerng.running:
    - image: registry/locksocket:{{ tag }}
    - port_bindings:
      - 9999:9999/tcp
    - environment:
{% for env, value in salt['pillar.get']('locksocket:envdir').items() %}
        {{ env }}: {{ value|string|yaml_encode }}
{% endfor %}
    - binds:
      - {{ chroot }}/config:/etc/locksocket:ro
    - watch:
      - file: config-file
      - dockerng: locksocket-image

```
docker.inspect_container <containername> --out=pprint
```
{'local': {u'AppArmorProfile': u'',
           u'Args': [],
           u'Config': {u'AttachStderr': True,
                       u'AttachStdin': False,
                       u'AttachStdout': True,
                       u'Cmd': [u'./runlocksocket.sh'],
                       u'Domainname': u'',
                       u'Entrypoint': None,
                       u'Env': [u'xxx=xxx',
                                u'xxxxx=xxxxx',
                                u'xxxx=xxx',
                                u'PATH=/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin',
                                u'LANG=C.UTF-8',
                                u'xxx=xxx',
                                u'PYTHON_VERSION=3.6.0',
                                u'PYTHON_PIP_VERSION=9.0.1'],
                       u'ExposedPorts': {u'9999/tcp': {}},
                       u'Hostname': u'd2f2ba9506c1',
                       u'Image': u'registry/locksocket:latest',
                       u'Labels': {},
                       u'OnBuild': None,
                       u'OpenStdin': False,
                       u'StdinOnce': False,
                       u'Tty': False,
                       u'User': u'',
                       u'Volumes': {u'/etc/locksocket': {}},
                       u'WorkingDir': u'/root/locksocket'},
           u'Created': u'2017-03-21T16:45:09.726972306Z',
           u'Driver': u'devicemapper',
           u'ExecIDs': None,
           u'GraphDriver': {u'Data': {u'DeviceId': u'539',
                                      u'DeviceName': u'docker-8:1-260452-adb10124edf28943636d55ffee378959577756f9e1d06c49b54da3a76ee5f84b',
                                      u'DeviceSize': u'10737418240'},
                            u'Name': u'devicemapper'},
           u'HostConfig': {u'AutoRemove': False,
                           u'Binds': [u'/srv/docker/locksocket/config:/etc/locksocket:ro'],
                           u'BlkioDeviceReadBps': None,
                           u'BlkioDeviceReadIOps': None,
                           u'BlkioDeviceWriteBps': None,
                           u'BlkioDeviceWriteIOps': None,
                           u'BlkioWeight': 0,
                           u'BlkioWeightDevice': None,
                           u'CapAdd': None,
                           u'CapDrop': None,
                           u'Cgroup': u'',
                           u'CgroupParent': u'',
                           u'ConsoleSize': [0, 0],
                           u'ContainerIDFile': u'',
                           u'CpuCount': 0,
                           u'CpuPercent': 0,
                           u'CpuPeriod': 0,
                           u'CpuQuota': 0,
                           u'CpuRealtimePeriod': 0,
                           u'CpuRealtimeRuntime': 0,
                           u'CpuShares': 0,
                           u'CpusetCpus': u'',
                           u'CpusetMems': u'',
                           u'Devices': None,
                           u'DiskQuota': 0,
                           u'Dns': None,
                           u'DnsOptions': None,
                           u'DnsSearch': None,
                           u'ExtraHosts': None,
                           u'GroupAdd': None,
                           u'IOMaximumBandwidth': 0,
                           u'IOMaximumIOps': 0,
                           u'IpcMode': u'',
                           u'Isolation': u'',
                           u'KernelMemory': 0,
                           u'Links': None,
                           u'LogConfig': {u'Config': {u'tag': u'{{ .ImageName }}/{{ .Name }}/{{ .ID }}/dev'},
                                          u'Type': u'journald'},
                           u'Memory': 0,
                           u'MemoryReservation': 0,
                           u'MemorySwap': 0,
                           u'MemorySwappiness': -1,
                           u'NanoCpus': 0,
                           u'NetworkMode': u'default',
                           u'OomKillDisable': False,
                           u'OomScoreAdj': 0,
                           u'PidMode': u'',
                           u'PidsLimit': 0,
                           u'PortBindings': {u'9999/tcp': [{u'HostIp': u'',
                                                            u'HostPort': u'9999'}]},
                           u'Privileged': False,
                           u'PublishAllPorts': False,
                           u'ReadonlyRootfs': False,
                           u'RestartPolicy': {u'MaximumRetryCount': 0,
                                              u'Name': u''},
                           u'Runtime': u'runc',
                           u'SecurityOpt': None,
                           u'ShmSize': 67108864,
                           u'UTSMode': u'',
                           u'Ulimits': None,
                           u'UsernsMode': u'',
                           u'VolumeDriver': u'',
                           u'VolumesFrom': None},
           u'HostnamePath': u'/var/lib/docker/containers/d2f2ba9506c1b51a336b073a416e58329122f52e8ad2755ead5e410d41b2fb0f/hostname',
           u'HostsPath': u'/var/lib/docker/containers/d2f2ba9506c1b51a336b073a416e58329122f52e8ad2755ead5e410d41b2fb0f/hosts',
           u'Id': u'd2f2ba9506c1b51a336b073a416e58329122f52e8ad2755ead5e410d41b2fb0f',
           u'Image': u'sha256:64a63c549c9c29d09d0d06e5f1e6f9efbfe9335d0b0ce5f195e746df472be96a',
           u'LogPath': u'',
           u'MountLabel': u'',
           u'Mounts': [{u'Destination': u'/etc/locksocket',
                        u'Mode': u'ro',
                        u'Propagation': u'',
                        u'RW': False,
                        u'Source': u'/srv/docker/locksocket/config',
                        u'Type': u'bind'}],
           u'Name': u'/locksocket',
           u'NetworkSettings': {u'Bridge': u'',
                                u'EndpointID': u'037e52e9d0edd02085de6484c205c942de0cf1ea501c70760a8bea012fa708ba',
                                u'Gateway': u'172.17.0.1',
                                u'GlobalIPv6Address': u'',
                                u'GlobalIPv6PrefixLen': 0,
                                u'HairpinMode': False,
                                u'IPAddress': u'172.17.0.6',
                                u'IPPrefixLen': 16,
                                u'IPv6Gateway': u'',
                                u'LinkLocalIPv6Address': u'',
                                u'LinkLocalIPv6PrefixLen': 0,
                                u'MacAddress': u'02:42:ac:11:00:06',
                                u'Networks': {u'bridge': {u'Aliases': None,
                                                          u'EndpointID': u'037e52e9d0edd02085de6484c205c942de0cf1ea501c70760a8bea012fa708ba',
                                                          u'Gateway': u'172.17.0.1',
                                                          u'GlobalIPv6Address': u'',
                                                          u'GlobalIPv6PrefixLen': 0,
                                                          u'IPAMConfig': None,
                                                          u'IPAddress': u'172.17.0.6',
                                                          u'IPPrefixLen': 16,
                                                          u'IPv6Gateway': u'',
                                                          u'Links': None,
                                                          u'MacAddress': u'02:42:ac:11:00:06',
                                                          u'NetworkID': u'01a56956cfea037f9f968f208a886e742bf3b5af271663811cb37ec7f3641b36'}},
                                u'Ports': {u'9999/tcp': [{u'HostIp': u'0.0.0.0',
                                                          u'HostPort': u'9999'}]},
                                u'SandboxID': u'8f80b0aad6ece10e2e150dfaebb427d9cba55b793529cac942438373ea534c5f',
                                u'SandboxKey': u'/var/run/docker/netns/8f80b0aad6ec',
                                u'SecondaryIPAddresses': None,
                                u'SecondaryIPv6Addresses': None},
           u'Path': u'./runlocksocket.sh',
           u'ProcessLabel': u'',
           u'ResolvConfPath': u'/var/lib/docker/containers/d2f2ba9506c1b51a336b073a416e58329122f52e8ad2755ead5e410d41b2fb0f/resolv.conf',
           u'RestartCount': 0,
           u'State': {u'Dead': False,
                      u'Error': u'',
                      u'ExitCode': 0,
                      u'FinishedAt': u'0001-01-01T00:00:00Z',
                      u'OOMKilled': False,
                      u'Paused': False,
                      u'Pid': 18514,
                      u'Restarting': False,
                      u'Running': True,
                      u'StartedAt': u'2017-03-21T16:45:11.345257162Z',
                      u'Status': u'running'}}}

```
```
>>> import docker
>>> docker.version_info
(1, 10, 6)
>>> 
```
---
`start` parameter

> Container exists and was running already, and was replaced. Should the new container be started?

yes, because it was replaced

> Container exists and was not running already, and was replaced. Should the new container be started?

yes, because it was replaced

> Container exists but was not replaced. It should not be started, correct?

correct if it has been started at least once

> Container did not previously exist. It should still be started even with start=False, because it is a new container, correct?

yes

`start=False` should be considered like regular `running` except we don't care if the container is not running after we started it.
---

ticosax commented:
> It turns out that the pre-existing code also did not have any support for stopped in the mod_watch

Damn, that's something I caught for the first time today. I thought it was a regression.

thx for fixing

terminalmage commented:
@ticosax OK, my guess is that the old code was passing empty strings for those three options, so they were set as empty strings in the container's HostConfig. Can you repeat the ``docker.inspect_container`` (with ``--out=pprint`` to make the data structure more readable) on one of the other containers you had previously managed with Salt prior to testing this PR, and check if those three params in the HostConfig are set to empty strings or set to ``None``?

terminalmage commented:
I made some fixes, squashed them, and rebased because there was another conflict due to a recent merge-forward @rallytime did.

Waiting to hear back from @ticosax on the most recent question that I asked him. If it is as I think it is, and the old code was passing empty strings for these values, then there's not much to do. That behavior should have been considered incorrect, and this PR would correct it. The best we can do is to document in the release notes and perhaps the salt/states/docker_container.py docstring, that when running the state for the first time in Nitrogen it may result in a replaced container with diffs similar to that.

ticosax commented:
> If it is as I think it is, and the old code was passing empty strings for these values, then there's not much to do.

@terminalmage That's the case, on a clean slate, there is no more wild diffs, you can ignore this issue.
let's get it merge.

terminalmage commented:
@ticosax good to hear. Before we get this merged I'm going to add a note about the diffs to the documentation. I'll get that done in the next couple hours and we can get this merged today. Thanks as always for your help!

terminalmage committed:
Add note about upgrading to new Docker Python bindings

terminalmage committed:
Add warning about possible replaced containers on first run after upgrade

terminalmage commented:
OK, warning has been added to the documentation, and I also added a warning to the state return data. The warning is even slightly different if it was triggered by ``test=True``:

```
# salt-call --local state.single docker_container.running foo image=arch command='sleep infinity' stop_timeout=5 test=True
local:
----------
          ID: foo
    Function: docker_container.running
      Result: None
     Comment: Container 'foo' would be replaced
     Started: 11:02:46.587626
    Duration: 1751.513 ms
     Changes:
              ----------
              container:
                  ----------
                  HostConfig:
                      ----------
                      DnsSearch:
                          ----------
                          new:
                              None
                          old:
    Warnings: The detected changes may be due to incorrect handling of arguments
              in earlier Salt releases. If this warning persists after running
              the state again without test=True, and no changes were made to the
              SLS file, then please report this.

Summary for local
------------
Succeeded: 1 (unchanged=1, changed=1)
Failed:    0
Warnings:  1
------------
Total states run:     1
Total run time:   1.752 s
# salt-call --local state.single docker_container.running foo image=arch command='sleep infinity' stop_timeout=5
local:
----------
          ID: foo
    Function: docker_container.running
      Result: True
     Comment: Replaced container 'foo'
     Started: 11:04:22.270419
    Duration: 7573.575 ms
     Changes:
              ----------
              container:
                  ----------
                  HostConfig:
                      ----------
                      DnsSearch:
                          ----------
                          new:
                              None
                          old:
              container_id:
                  ----------
                  added:
                      396399ac2f2fcb36aac08a06bf51db50e8a4d03d825bf84d57c875a0444b6302
                  removed:
                      - b0ebba05b365ed15e37bf06df53df417aed10a12a1f7f176be35336e280aea3c
    Warnings: The detected changes may be due to incorrect handling of arguments
              in earlier Salt releases. If this warning persists after running
              the state again, and no changes were made to the SLS file, then
              please report this.

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
Warnings:  1
------------
Total states run:     1
Total run time:   7.574 s
# salt-call --local state.single docker_container.running foo image=arch command='sleep infinity' stop_timeout=5
local:
----------
          ID: foo
    Function: docker_container.running
      Result: True
     Comment: Container 'foo' is already configured as specified
     Started: 11:04:42.164054
    Duration: 1724.74 ms
     Changes:

Summary for local
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:   1.725 s
```

Just waiting on tests

terminalmage committed:
Check container diff for evidence of improper arg handling in earlier releases

Also fix an issue with the result was not set to None when running the
state in test mode.
